### PR TITLE
[SPARK-10614][CORE] Change SystemClock to derive time from System.nanoTime() instead of System.currentTimeMillis()

### PR DESCRIPTION
**Pull request by markgrover**
## What changes were proposed in this pull request?

Change SystemClock to derive time from System.nanoTime() instead of System.currentTimeMillis(). This is because using System.currentTimeMillis() makes waitTillTime() routine brittle against systems like VMs where time can go forward and backwards.

Shout out to @steveloughran for doing a good chunk of research and a previous PR (#8766) on this in 2015.
## How was this patch tested?

Build is green, I am running unit tests, they are looking green, but are also taking a while. So, filing this PR a little proactively.

### CODE REVIEW

**Comment by vanzin** [comment]
Can you change the PR title to reflect the actual change, and not the bug's title?

File: core/src/main/scala/org/apache/spark/util/Clock.scala [hunk 1]
```diff
@@ -21,7 +21,14 @@ package org.apache.spark.util
  * An interface to represent clocks, so that they can be mocked out in unit tests.
  */
 private[spark] trait Clock {
+  /** @return the time in milliseconds */
   def getTimeMillis(): Long
+
+  /**
+   * Wait until a clock's view of current time reaches the specified target time.
+   * @param targetTime block until the current time is at least this value
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: nit: blank line before argument list

File: core/src/main/scala/org/apache/spark/util/Clock.scala [hunk 2]
```diff
@@ -33,18 +40,18 @@ private[spark] class SystemClock extends Clock {
   val minPollTime = 25L
 
   /**
-   * @return the same time (milliseconds since the epoch)
-   *         as is reported by `System.currentTimeMillis()`
+   * @return the same time in milliseconds as is derived from `System.nanoTime()`
    */
-  def getTimeMillis(): Long = System.currentTimeMillis()
+  def getTimeMillis(): Long = System.nanoTime() / (1000 * 1000)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: `TimeUnit.NANOSECONDS.toMillis(...)`

File: core/src/main/scala/org/apache/spark/util/Clock.scala (hunk 1 above)
Comment (markgrover) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: Will do

File: core/src/main/scala/org/apache/spark/util/Clock.scala (hunk 2 above)
Comment (markgrover) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: Will do

**Comment by markgrover** [comment]
The unit test failures from the previous run looked unrelated, fyi.

**Comment by vanzin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/util/Clock.scala [hunk 3]
```diff
@@ -17,11 +17,21 @@
...
+  /** @return the time in milliseconds */
   def getTimeMillis(): Long
+
+  /**
+   * Wait until a clock's view of current time reaches the specified target time.
+   *
+   * @param targetTime block until the current time is at least this value
+   * @return clock's view of current time when the wait has completed
+   */
   def waitTillTime(targetTime: Long): Long
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: what's the unit?

Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: it'd be good to rename the variable to reflect the unit, similar to getTimeMillis

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
@markgrover these two tests fail consistently just on your PR, so it looks like they're not as unrelated as you thought.

**Comment by markgrover** [comment]
Thanks @vanzin Yeah, I noticed that too. Will take a look.

**Comment by markgrover** [comment]
And, thanks for your feedback, @rxin, I will take a look.

**Comment by zsxwing** [comment]
> This method can only be used to measure elapsed time and is not related to any other notion of system or wall-clock time. The value returned represents nanoseconds since some fixed but arbitrary origin time (perhaps in the future, so values may be negative). The same origin is used by all invocations of this method in an instance of a Java virtual machine; other virtual machine instances are likely to use a different origin.

As per javadoc, `System.nanoTime` cannot be used to get the current system time. I saw most of places just use `getTimeMillis` to get the current system time and display it on UI. I don't think we can change it to `System.nanoTime`.

@markgrover  could you close this pr, please?

**Comment by vanzin** [comment]
I don't think closing this PR is the right thing to do. This solves an actual problem, in that we have code that depends on the clock being monotonic, but a non-monotonic clock is being used.

If there's a need to differentiate that from actual system time, then perhaps we need to extend the `Clock` API. But there's an actual issue that this change is fixing.

**Comment by zsxwing** [comment]
@vanzin I didn't mean closing the JIRA ticket. Let's just close this stale PR since it doesn't have the correct fix. We can always submit a new PR if necessary.

**Pull request closed without merging**